File: apps/worker/src/aviary_worker/crypto.py
import base64
import hashlib
from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def _derive_key(raw: str) -> bytes:
    if len(raw) == 64:
        try:
            return bytes.fromhex(raw)
        except ValueError:
            pass

    utf = raw.encode("utf-8")
    if len(utf) == 32:
        return utf

    return hashlib.sha256(utf).digest()


def decrypt_secret(payload: str, raw_key: str) -> str:
    iv_b64, tag_b64, encrypted_b64 = payload.split(":")
    iv = base64.b64decode(iv_b64)
    tag = base64.b64decode(tag_b64)
    encrypted = base64.b64decode(encrypted_b64)

    aesgcm = AESGCM(_derive_key(raw_key))
    plaintext = aesgcm.decrypt(iv, encrypted + tag, None)
    return plaintext.decode("utf-8")

File: apps/api/src/crypto.ts
import { createCipheriv, createDecipheriv, createHash, randomBytes } from "node:crypto";

export function getKey(raw: string): Buffer {
  if (/^[0-9a-fA-F]{64}$/.test(raw)) {
    return Buffer.from(raw, "hex");
  }

  const buf = Buffer.from(raw, "utf8");
  if (buf.length === 32) {
    return buf;
  }

  return createHash("sha256").update(raw).digest();
}

export function encryptSecret(plainText: string, key: Buffer): string {
  const iv = randomBytes(12);
  const cipher = createCipheriv("aes-256-gcm", key, iv);
  const encrypted = Buffer.concat([cipher.update(plainText, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag();
  return `${iv.toString("base64")}:${tag.toString("base64")}:${encrypted.toString("base64")}`;
}

export function decryptSecret(payload: string, key: Buffer): string {
  const [ivB64, tagB64, encryptedB64] = payload.split(":");
  if (!ivB64 || !tagB64 || !encryptedB64) {
    throw new Error("Invalid encrypted payload");
  }

  const decipher = createDecipheriv("aes-256-gcm", key, Buffer.from(ivB64, "base64"));
  decipher.setAuthTag(Buffer.from(tagB64, "base64"));
  const plain = Buffer.concat([
    decipher.update(Buffer.from(encryptedB64, "base64")),
    decipher.final()
  ]);

  return plain.toString("utf8");
}

export function sha256(value: string): string {
  return createHash("sha256").update(value).digest("hex");
}
